autopush/router/apns2.py

```diff
@@ -0,0 +1,152 @@
+        Access** application. To extract the proper PEM formatted data, you
+        can use the following commands:
+
+        ```
+        openssl pkcs12 -in file.p12 -out apns_cert.pem -clcerts -nokeys
+        openssl pkcs12 -in file.p12 -out apns_key.pem -nocerts -nodes
+        ```
+
+        The *topic* is the Bundle ID of the bridge recipient iOS application.
+        Since the cert needs to be tied directly to an application, the topic
+        is usually similar to "com.example.MyApplication".
+
+        :param cert_file: Path to the PEM formatted APNs certification file.
+        :type cert_file: str
+        :param key_file: Path to the PEM formatted APNs key file.
+        :type key_file: str
+        :param topic: The *Bundle ID* that identifies the assoc. iOS app.
+        :type topic: str
+        :param alt: Use the alternate APNs publication port (if 443 is blocked)
+        :type alt: bool
+        :param use_sandbox: Use the development sandbox
+        :type alt: bool
+        :param max_connections: Max number of pooled connections to use
+        :type max_connections: int
+        :param min_connections: Min number of pooled connections to use
+        :type min_connections: int
+        :param logger: Status logger
+        :type logger: logger
+        :param metrics: Metric recorder
+        :type metrics: autopush.metrics.IMetric
+        """
+        self.server = SERVER if not use_sandbox else SANDBOX
+        self.port = 443 if not alt else 2197
+        self.ssl_context = hyper.tls.init_context(cert=(cert_file, key_file))
+        self.log = logger
+        self.metrics = metrics
+        self.topic = topic
+        self._max_connections = max_connections
+        self._min_connections = min_connections
+        self.connections = deque(maxlen=max_connections)
+        if self.log:
+            self.log.info("Starting APNS connection")
+        # HTTP20Connection is not thread safe, so it's best to keep processing
+        # of the result in the same thread.
+        for i in xrange(0, min_connections):
+            self._get_connection()
+
+    def send(self, router_token, payload, apns_id=None,
+             priority=True, topic=None, exp=None):
+        """Send the dict of values to the remote bridge
+
+        This sends the raw data to the remote bridge application using the
+        APNS2 HTTP2 API.
+
+        :param router_token: APNs provided hex token identifying recipient
+        :type router_token: str
+        :param payload: Data to send to recipient
+        :type payload: dict
+        :param priority: True is high priority, false is low priority
+        :type priority: bool
+        :param topic: BundleID for the recipient application (overides default)
+        :type topic: str
+        :param exp: Message expiration timestamp
+        :type exp: timestamp
+
+        """
+        if not apns_id:
+            apns_id = str(uuid.uuid4()).lower()
```

Comment:
I can't imagine Python would ever ever change this, as that would probably break the world.